File: src/test/java/cz/vhromada/catalog/web/common/MovieUtils.java
package cz.vhromada.catalog.web.common;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.SoftAssertions.assertSoftly;

import java.util.List;

import cz.vhromada.catalog.common.Language;
import cz.vhromada.catalog.entity.Medium;
import cz.vhromada.catalog.entity.Movie;
import cz.vhromada.catalog.utils.CollectionUtils;
import cz.vhromada.catalog.web.fo.MovieFO;
import cz.vhromada.catalog.web.fo.TimeFO;

/**
 * A class represents utility class for movies.
 *
 * @author Vladimir Hromada
 */
public final class MovieUtils {

    /**
     * Creates a new instance of MovieUtils.
     */
    private MovieUtils() {
    }

    /**
     * Returns FO for movie.
     *
     * @return FO for movie
     */
    public static MovieFO getMovieFO() {
        final MovieFO movie = new MovieFO();
        movie.setId(CatalogUtils.ID);
        movie.setCzechName("czName");
        movie.setOriginalName("origName");
        movie.setYear(Integer.toString(CatalogUtils.YEAR));
        movie.setLanguage(Language.EN);
        movie.setSubtitles(CollectionUtils.newList(Language.CZ));
        movie.setCsfd("Csfd");
        movie.setImdb(true);
        movie.setImdbCode("1000");
        movie.setWikiEn(CatalogUtils.EN_WIKI);
        movie.setWikiCz(CatalogUtils.CZ_WIKI);
        movie.setPicture(CatalogUtils.ID);
        movie.setNote(CatalogUtils.NOTE);
        movie.setPosition(CatalogUtils.POSITION);
        movie.setMedia(CollectionUtils.newList(TimeUtils.getTimeFO()));
        movie.setGenres(CollectionUtils.newList(CatalogUtils.ID));

        return movie;
    }

    /**
     * Returns movie.
     *
     * @return movie
     */
    public static Movie getMovie() {
        final Medium medium = new Medium();
        medium.setId(CatalogUtils.ID);
        medium.setNumber(CatalogUtils.NUMBER);
        medium.setLength(CatalogUtils.LENGTH);

        final Movie movie = new Movie();
        movie.setId(CatalogUtils.ID);
        movie.setCzechName("czName");
        movie.setOriginalName("origName");
        movie.setYear(CatalogUtils.YEAR);
        movie.setLanguage(Language.EN);
        movie.setSubtitles(CollectionUtils.newList(Language.CZ));
        movie.setCsfd("Csfd");
        movie.setImdbCode(1000);
        movie.setWikiEn(CatalogUtils.EN_WIKI);
        movie.setWikiCz(CatalogUtils.CZ_WIKI);
        movie.setPicture(CatalogUtils.ID);
        movie.setNote(CatalogUtils.NOTE);
        movie.setPosition(CatalogUtils.POSITION);
        movie.setMedia(CollectionUtils.newList(medium));
        movie.setGenres(CollectionUtils.newList(GenreUtils.getGenre()));

        return movie;
    }

    /**
     * Asserts movie deep equals.
     *
     * @param expected expected FO for movie
     * @param actual   actual movie
     */
    public static void assertMovieDeepEquals(final MovieFO expected, final Movie actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertSoftly(softly -> {
            softly.assertThat(actual.getId()).isEqualTo(expected.getId());
            softly.assertThat(actual.getCzechName()).isEqualTo(expected.getCzechName());
            softly.assertThat(actual.getOriginalName()).isEqualTo(expected.getOriginalName());
            softly.assertThat(actual.getYear()).isEqualTo(Integer.parseInt(expected.getYear()));
            softly.assertThat(actual.getLanguage()).isEqualTo(expected.getLanguage());
            softly.assertThat(actual.getSubtitles()).isEqualTo(expected.getSubtitles());
            softly.assertThat(actual.getCsfd()).isEqualTo(expected.getCsfd());
            CatalogUtils.assertImdbCodeDeepEquals(expected.getImdb(), expected.getImdbCode(), actual.getImdbCode());
            softly.assertThat(actual.getWikiEn()).isEqualTo(expected.getWikiEn());
            softly.assertThat(actual.getWikiCz()).isEqualTo(expected.getWikiCz());
            softly.assertThat(actual.getPicture()).isEqualTo(expected.getPicture());
            softly.assertThat(actual.getNote()).isEqualTo(expected.getNote());
            softly.assertThat(actual.getPosition()).isEqualTo(expected.getPosition());
            assertMediaDeepEquals(expected.getMedia(), actual.getMedia());
            GenreUtils.assertGenresDeepEquals(expected.getGenres(), actual.getGenres());
        });
    }

    /**
     * Asserts media deep equals.
     *
     * @param expected expected list of FO for time
     * @param actual   actual list of medium
     */
    public static void assertMediaDeepEquals(final List<TimeFO> expected, final List<Medium> actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertThat(actual.size()).isEqualTo(expected.size());
        for (int i = 0; i < expected.size(); i++) {
            assertMediumDeepEquals(expected.get(i), actual.get(i), i);
        }
    }

    /**
     * Asserts medium deep equals.
     *
     * @param expected expected FO for time
     * @param actual   actual medium
     * @param index    index
     */
    public static void assertMediumDeepEquals(final TimeFO expected, final Medium actual, final int index) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertSoftly(softly -> {
            softly.assertThat(actual.getId()).isNull();
            softly.assertThat(actual.getNumber()).isEqualTo(index + 1);
            TimeUtils.assertTimeDeepEquals(expected, actual.getLength());
        });
    }

    /**
     * Asserts movie deep equals.
     *
     * @param expected expected movie
     * @param actual   actual FO for movie
     */
    public static void assertMovieDeepEquals(final Movie expected, final MovieFO actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertSoftly(softly -> {
            softly.assertThat(actual.getId()).isEqualTo(expected.getId());
            softly.assertThat(actual.getCzechName()).isEqualTo(expected.getCzechName());
            softly.assertThat(actual.getOriginalName()).isEqualTo(expected.getOriginalName());
            softly.assertThat(actual.getYear()).isEqualTo(Integer.toString(expected.getYear()));
            softly.assertThat(actual.getLanguage()).isEqualTo(expected.getLanguage());
            softly.assertThat(actual.getSubtitles()).isEqualTo(expected.getSubtitles());
            softly.assertThat(actual.getCsfd()).isEqualTo(expected.getCsfd());
            CatalogUtils.assertImdbDeepEquals(expected.getImdbCode(), actual.getImdb(), actual.getImdbCode());
            softly.assertThat(actual.getWikiEn()).isEqualTo(expected.getWikiEn());
            softly.assertThat(actual.getWikiCz()).isEqualTo(expected.getWikiCz());
            softly.assertThat(actual.getPicture()).isEqualTo(expected.getPicture());
            softly.assertThat(actual.getNote()).isEqualTo(expected.getNote());
            softly.assertThat(actual.getPosition()).isEqualTo(expected.getPosition());
            assertMediumListDeepEquals(expected.getMedia(), actual.getMedia());
            GenreUtils.assertGenreListDeepEquals(expected.getGenres(), actual.getGenres());
        });
    }

    /**
     * Asserts media deep equals.
     *
     * @param expected expected list of medium
     * @param actual   actual list of FO for time
     */
    public static void assertMediumListDeepEquals(final List<Medium> expected, final List<TimeFO> actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertThat(actual.size()).isEqualTo(expected.size());
        for (int i = 0; i < expected.size(); i++) {
            final TimeFO actualTime = actual.get(i);
            final int index = i;
            assertSoftly(softly -> {
                softly.assertThat(actualTime).isNotNull();
                TimeUtils.assertTimeDeepEquals(actualTime, expected.get(index).getLength());
            });
        }
    }

}

File: src/test/kotlin/cz/vhromada/catalog/web/common/MovieUtils.kt
package cz.vhromada.catalog.web.common

import cz.vhromada.catalog.entity.Medium
import cz.vhromada.catalog.entity.Movie
import cz.vhromada.catalog.web.fo.MovieFO
import cz.vhromada.catalog.web.fo.TimeFO
import cz.vhromada.common.entity.Language
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.SoftAssertions.assertSoftly

/**
 * A class represents utility class for movies.
 *
 * @author Vladimir Hromada
 */
object MovieUtils {

    /**
     * Returns FO for movie.
     *
     * @return FO for movie
     */
    fun getMovieFO(): MovieFO {
        return MovieFO(id = CatalogUtils.ID,
                czechName = "czName",
                originalName = "origName",
                year = CatalogUtils.YEAR.toString(),
                language = Language.EN,
                subtitles = listOf(Language.CZ),
                csfd = "Csfd",
                imdb = true,
                imdbCode = "1000",
                wikiEn = CatalogUtils.EN_WIKI,
                wikiCz = CatalogUtils.CZ_WIKI,
                picture = CatalogUtils.ID,
                note = CatalogUtils.NOTE,
                position = CatalogUtils.POSITION,
                media = listOf(TimeUtils.getTimeFO()),
                genres = listOf(CatalogUtils.ID))
    }

    /**
     * Returns movie.
     *
     * @return movie
     */
    fun getMovie(): Movie {
        val medium = Medium(id = CatalogUtils.ID,
                number = CatalogUtils.NUMBER,
                length = CatalogUtils.LENGTH)

        return Movie(id = CatalogUtils.ID,
                czechName = "czName",
                originalName = "origName",
                year = CatalogUtils.YEAR,
                language = Language.EN,
                subtitles = listOf(Language.CZ),
                csfd = "Csfd",
                imdbCode = 1000,
                wikiEn = CatalogUtils.EN_WIKI,
                wikiCz = CatalogUtils.CZ_WIKI,
                picture = CatalogUtils.ID,
                note = CatalogUtils.NOTE,
                position = CatalogUtils.POSITION,
                media = listOf(medium),
                genres = listOf(GenreUtils.getGenre()))
    }

    /**
     * Asserts movie deep equals.
     *
     * @param expected expected FO for movie
     * @param actual   actual movie
     */
    fun assertMovieDeepEquals(expected: MovieFO?, actual: Movie?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertSoftly {
            it.assertThat(actual!!.id).isEqualTo(expected!!.id)
            it.assertThat(actual.czechName).isEqualTo(expected.czechName)
            it.assertThat(actual.originalName).isEqualTo(expected.originalName)
            it.assertThat(actual.year).isEqualTo(Integer.parseInt(expected.year))
            it.assertThat<Language>(actual.language).isEqualTo(expected.language)
            it.assertThat<Language>(actual.subtitles).isEqualTo(expected.subtitles)
            it.assertThat(actual.csfd).isEqualTo(expected.csfd)
            CatalogUtils.assertImdbCodeDeepEquals(expected.imdb, expected.imdbCode, actual.imdbCode)
            it.assertThat(actual.wikiEn).isEqualTo(expected.wikiEn)
            it.assertThat(actual.wikiCz).isEqualTo(expected.wikiCz)
            it.assertThat(actual.picture).isEqualTo(expected.picture)
            it.assertThat(actual.note).isEqualTo(expected.note)
            it.assertThat(actual.position).isEqualTo(expected.position)
            assertMediaDeepEquals(expected.media, actual.media)
            GenreUtils.assertGenresDeepEquals(expected.genres, actual.genres)
        }
    }

    /**
     * Asserts media deep equals.
     *
     * @param expected expected list of FO for time
     * @param actual   actual list of medium
     */
    private fun assertMediaDeepEquals(expected: List<TimeFO?>?, actual: List<Medium?>?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertThat(actual!!.size).isEqualTo(expected!!.size)
        for (i in expected.indices) {
            assertMediumDeepEquals(expected[i], actual[i], i)
        }
    }

    /**
     * Asserts medium deep equals.
     *
     * @param expected expected FO for time
     * @param actual   actual medium
     * @param index    index
     */
    private fun assertMediumDeepEquals(expected: TimeFO?, actual: Medium?, index: Int) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertSoftly {
            it.assertThat(actual!!.id).isNull()
            it.assertThat(actual.number).isEqualTo(index + 1)
            TimeUtils.assertTimeDeepEquals(expected, actual.length)
        }
    }

    /**
     * Asserts movie deep equals.
     *
     * @param expected expected movie
     * @param actual   actual FO for movie
     */
    fun assertMovieDeepEquals(expected: Movie?, actual: MovieFO?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertSoftly {
            it.assertThat(actual!!.id).isEqualTo(expected!!.id)
            it.assertThat(actual.czechName).isEqualTo(expected.czechName)
            it.assertThat(actual.originalName).isEqualTo(expected.originalName)
            it.assertThat(actual.year).isEqualTo(expected.year.toString())
            it.assertThat<Language>(actual.language).isEqualTo(expected.language)
            it.assertThat(actual.subtitles).isEqualTo(expected.subtitles)
            it.assertThat(actual.csfd).isEqualTo(expected.csfd)
            CatalogUtils.assertImdbDeepEquals(expected.imdbCode, actual.imdb, actual.imdbCode)
            it.assertThat(actual.wikiEn).isEqualTo(expected.wikiEn)
            it.assertThat(actual.wikiCz).isEqualTo(expected.wikiCz)
            it.assertThat(actual.picture).isEqualTo(expected.picture)
            it.assertThat(actual.note).isEqualTo(expected.note)
            it.assertThat(actual.position).isEqualTo(expected.position)
            assertMediumListDeepEquals(expected.media, actual.media)
            GenreUtils.assertGenreListDeepEquals(expected.genres, actual.genres)
        }
    }

    /**
     * Asserts media deep equals.
     *
     * @param expected expected list of medium
     * @param actual   actual list of FO for time
     */
    private fun assertMediumListDeepEquals(expected: List<Medium?>?, actual: List<TimeFO?>?) {
        assertSoftly {
            it.assertThat(expected).isNotNull
            it.assertThat(actual).isNotNull
        }
        assertThat(actual!!.size).isEqualTo(expected!!.size)
        for (i in expected.indices) {
            val actualTime = actual[i]
            assertSoftly {
                it.assertThat(actualTime).isNotNull
                TimeUtils.assertTimeDeepEquals(actualTime, expected[i]!!.length)
            }
        }
    }

}
